project-koku/koku
# [COST-1026] Cost 1026 OCPGCP Processing take 2

Fixes [COST-1026](https://issues.redhat.com/browse/COST-1026)

Changes proposed in this PR:
*This PR is the backend processing for OCP on GCP to prepare for the APIs.
*
SMOKES RESULTS: [GCP Smokes](https://ci.ext.devshift.net/job/project-koku-koku-pr-check/3729/)
OCP Trino Smokes: ran into a timeout in ephemeral but no actual issues, Luke ran them locally and they passed, see his comment below.

There are two sets of testing instructions, one of them is extended instructions for if you are really feeling like checking stuff out (which i do encourage).
Testing Instructions:
Since this PR only includes the processing work and not the api work, testing is based on info loaded in to the db. A new ocp and gcp source has been added to the make create-test-customer and make load-test-customer-data commands that will load in an ocp source and a matching gcp source. Once loaded in you should be able to check the tables and see relevant entries in the tables. You can also independently create nise data and send an ocp source then a gcp source to populate the data.
1. Run `make create-test-customer-data`
2. Ingest this data `load-test-customer-data`
3. Go into the db and check the tables for entires in the following tables:
```
reporting_ocpgcp_costentrylineitem_daily_summary
reporting_ocpgcpcostentrylineitem_project_daily_summary
Tags should be in: reporting_ocpgcptags_summary and reporting_ocpgcptags_values
reporting_ocpgcp_cost_summary_p
reporting_ocpgcp_cost_summary_by_service_p
reporting_ocpgcp_cost_summary_region_p
reporting_ocpgcp_cost_summary_gcp_project_p
reporting_ocpgcp_cost_summary_account_p
reporting_ocpgcp_compute_summary_p
reporting_ocpgcp_database_summary_p
reporting_ocpgcp_network_summary_p
reporting_ocpgcp_storage_summary_p
```
---
EXTENDED TESTING INSTRUCTIONS (will involve some math and local IQE plugin):

1. Bring everything up in trino mode, such as `make docker-up-min-trino`
2. Checkout the `COST-1029-enable-ocp-gcp` branch of the cost iqe plugin
3. Run the following command to trigger the ocp on gcp raw calc test however you run your iqe tests: `ENV_FOR_DYNACONF=local iqe tests plugin cost_management -k test_api_ocp_on_gcp_source_raw_calc --pdb`
4. Wait for data ingestion, eventually the test will fail and pause because the api doesnt exist but you will still have a db of processed data.
5. Run some queries on the postgres db to compare gcp to ocp on gcp results

Math behind the numbers (at the time of writing): Each gcp yaml has a cost of 2 per hour with 4 generators, compute, storage, network, database. With 2 * 24 hours * 58 full days = 2784 + 13 hours today * 2 = 2810 for each generator. a total of 11240 combined.
- Comparing unblended costs: 
```
postgres=# select sum(unblended_cost) from reporting_gcpcostentrylineitem_daily_summary;
       sum       
-----------------
 11240.000000000

postgres=# select sum(unblended_cost) from reporting_ocpgcpcostlineitem_daily_summary_p;
       sum       
-----------------
 11240.000000000

postgres=# select sum(unblended_cost) from reporting_ocpgcpcostlineitem_project_daily_summary_p;
          sum          
-----------------------
 11240.000000000000000
```

- Comparing credits (200 per month for the compute generator and 100 per month from storage for 2 months so 600 total):
```
postgres=# select sum(credit_amount) from reporting_gcpcostentrylineitem_daily_summary;
      sum       
----------------
 -600.000000036
                             
postgres=# select sum(credit_amount) from reporting_ocpgcpcostlineitem_daily_summary_p;
      sum       
----------------
 -599.999999987

postgres=# select sum(credit_amount) from reporting_gcp_storage_summary_p;
      sum       
----------------
 -200.000000012

postgres=# select sum(credit_amount) from reporting_ocpgcp_storage_summary_p;
      sum       
----------------
 -199.999999994

postgres=# select sum(credit_amount) from reporting_gcp_network_summary_p;
     sum     
-------------
 0.000000000

postgres=# select sum(credit_amount) from reporting_ocpgcp_network_summary_p;
     sum     
-------------
 0.000000000
```

- Comparing Storage
```
postgres=# select sum(unblended_cost) from reporting_gcp_storage_summary_p;
      sum       
----------------
 2810.000000000

postgres=# select sum(unblended_cost) from reporting_ocpgcp_storage_summary_p;
      sum       
----------------
 2810.000000000
```

- Comparing Networking
```
postgres=# select sum(unblended_cost) from reporting_gcp_network_summary_p;
      sum       
----------------
 2810.000000000

postgres=# select sum(unblended_cost) from reporting_ocpgcp_network_summary_p;
      sum       
----------------
 2810.000000000
```

---
Additional Context:
OCP on GCP is not matched by resource ID as our other cloud providers are, but based on a specific tag of {"kubernetes-io-cluster-{cluster-id}": "owned"} or our special case tags.
